### PR TITLE
Port

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

File: examples_tests/20.Megatexture/main.cpp [hunk 1]
```diff
@@ -521,7 +522,7 @@ APP_CONSTRUCTOR(MeshLoadersApp)
             vt->commit(cm.addr, cm.texture.get(), cm.subresource, cm.uwrap, cm.vwrap, cm.border);
         }
 
-        auto gpuvt = core::make_smart_refctd_ptr<video::IGPUVirtualTexture>(engine->logicalDevice.get(), engine->queues, vt.get());
+        auto gpuvt = core::make_smart_refctd_ptr<video::IGPUVirtualTexture>(engine->logicalDevice.get(), engine->gpuTransferFence.get(), engine->queues[CommonAPI::InitOutput<1>::EQT_TRANSFER_UP], vt.get());
```
Comment (devshgraphicsprogramming) [inline examples_tests/20.Megatexture/main.cpp:525]: fence and queue cannot go into the constructor @Przemog1

Comment (devshgraphicsprogramming) [reply examples_tests/20.Megatexture/main.cpp:525]: ah ok its the constructor...

you should just take the queue, and make your own fence in there.

File: include/nbl/video/utilities/IGPUVirtualTexture.h [hunk 2]
```diff
@@ -38,24 +38,25 @@ class IGPUVirtualTexture final : public asset::IVirtualTexture<IGPUImageView, IG
         uint32_t& waitSemaphoreCount, IGPUSemaphore* const* &semaphoresToWaitBeforeExecution, const asset::E_PIPELINE_STAGE_FLAGS* &stagesToWaitForPerSemaphore
     )
     {
+        core::smart_refctd_ptr<IUtilities> utilities = core::make_smart_refctd_ptr<IUtilities>(core::smart_refctd_ptr<ILogicalDevice>(logicalDevice));
```
Comment (devshgraphicsprogramming) [inline include/nbl/video/utilities/IGPUVirtualTexture.h:41]: dont make a new utilitiy class, take the existing created one

File: include/nbl/video/utilities/IGPUVirtualTexture.h [hunk 3]
```diff
@@ -80,7 +81,7 @@ class IGPUVirtualTexture final : public asset::IVirtualTexture<IGPUImageView, IG
             auto* cpuStorage = static_cast<asset::ICPUVirtualTexture::ICPUVTResidentStorage*>(pair.second.get());
             const asset::E_FORMAT_CLASS fmtClass = pair.first;
 
-            m_storage.insert({fmtClass,core::make_smart_refctd_ptr<IGPUVTResidentStorage>(m_logicalDevice.get(),queue.get(),cpuStorage)});
+            m_storage.insert({fmtClass,core::make_smart_refctd_ptr<IGPUVTResidentStorage>(m_logicalDevice.get(),cmdbuf,fenceToSignal,queue,cpuStorage,waitSemaphoreCount,semaphoresToWaitBeforeExecution,stagesToWaitForPerSemaphore)});
```
Comment (devshgraphicsprogramming) [inline include/nbl/video/utilities/IGPUVirtualTexture.h:84]: what's going on here?

File: include/nbl/video/utilities/IGPUVirtualTexture.h [hunk 4]
```diff
@@ -38,24 +38,25 @@ class IGPUVirtualTexture final : public asset::IVirtualTexture<IGPUImageView, IG
...
-            logicalDevice->updateBufferRangeViaStagingBuffer(
+            utilities->updateBufferRangeViaStagingBuffer(
                 cmdbuf,fence,queue,asset::SBufferRange<IGPUBuffer>{0u,gpuTexelBuffer->getSize(),gpuTexelBuffer},_cpuimg->getBuffer()->getPointer(),
                 waitSemaphoreCount,semaphoresToWaitBeforeExecution,stagesToWaitForPerSemaphore
             );
 
             auto regions = _cpuimg->getRegions();
             assert(regions.size());
-            gpuImage = logicalDevice->createFilledDeviceLocalGPUImageOnDedMem(cmdbuf,std::move(cpuImageParams),gpuTexelBuffer.get(),regions.size(),regions.begin());
+            gpuImage = utilities->createFilledDeviceLocalGPUImageOnDedMem(cmdbuf,std::move(cpuImageParams),gpuTexelBuffer.get(),regions.size(),regions.begin());
```
Comment (devshgraphicsprogramming) [inline include/nbl/video/utilities/IGPUVirtualTexture.h:59]: you actually need to submit the commandbuffer while signalling the fence later on, are you doing that?

File: include/nbl/asset/ICPUImage.h [hunk 5]
```diff
@@ -94,6 +94,7 @@ class ICPUImage final : public IImage, public IAsset
 			assert(false); //TODO: the stuff below causes a compile error
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUImage.h:94]: this needs to be fixed, urgently

File: include/nbl/asset/ICPUImage.h [hunk 6]
```diff
@@ -94,6 +94,7 @@ class ICPUImage final : public IImage, public IAsset
...
 			return nullptr; //core::smart_refctd_ptr<immutable_refctd_array_t>(reinterpret_cast<immutable_refctd_array_t*>(regions.get()));
 		}
+		inline auto getRegionArray() { return regions; }
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUImage.h:97]: we shouldn't have a non-const getter for these

File: examples_tests/03.GPU_Mesh/main.cpp [hunk 7]
```diff
@@ -277,86 +276,86 @@ APP_CONSTRUCTOR(GPUMesh)
...
-		//	}
-
-		//	driver->bindGraphicsPipeline(mb->getPipeline());
-		//	driver->pushConstants(mb->getPipeline()->getLayout(), asset::ISpecializedShader::ESS_VERTEX, 0u, sizeof(core::matrix4SIMD), mvp.pointer());
-		//	driver->drawMeshBuffer(mb.get());
-
-		//	upStreamBuff->multi_free(2u, (uint32_t*)&offsets, (uint32_t*)&sizes, driver->placeFence());
-		//}
-		//driver->endScene();
+			//	auto upStreamBuff = driver->getDefaultUpStreamingBuffer();
```
Comment (Przemog1) [inline examples_tests/03.GPU_Mesh/main.cpp:279]: TODO

File: src/nbl/asset/utils/CShaderIntrospector.cpp [hunk 8]
```diff
@@ -325,6 +326,8 @@ core::smart_refctd_ptr<ICPUDescriptorSetLayout> CShaderIntrospector::createAppro
         binding.count = introBnd.descriptorCount;
         if (introBnd.descCountIsSpecConstant)
         {
+            //TODO: delete this line:
+            refShader = *shaders.begin();
```
Comment (Przemog1) [inline src/nbl/asset/utils/CShaderIntrospector.cpp:330]: !

File: examples_tests/05.NablaTutorialExample/main.cpp [hunk 9]
```diff
@@ -199,8 +234,10 @@ class NablaTutorialExampleApp : public ApplicationBase
 		auto cpuVertexShader = core::smart_refctd_ptr_static_cast<ICPUSpecializedShader>(assetManager->findAssets("nbl/builtin/material/lambertian/singletexture/specialized_shader.vert", types)->front().getContents().begin()[0]);
 		auto cpuFragmentShader = core::smart_refctd_ptr_static_cast<ICPUSpecializedShader>(assetManager->findAssets("nbl/builtin/material/lambertian/singletexture/specialized_shader.frag", types)->front().getContents().begin()[0]);
 
+		cpu2gpuParams.beginCommandBuffers();
 		auto gpuVertexShader = cpu2gpu.getGPUObjectsFromAssets(&cpuVertexShader.get(), &cpuVertexShader.get() + 1, cpu2gpuParams)->front();
 		cpu2gpuParams.waitForCreationToComplete();
+		cpu2gpuParams.beginCommandBuffers();
 		auto gpuFragmentShader = cpu2gpu.getGPUObjectsFromAssets(&cpuFragmentShader.get(), &cpuFragmentShader.get() + 1, cpu2gpuParams)->front();
 		cpu2gpuParams.waitForCreationToComplete();
 		cpu2gpuParams.waitForCreationToComplete();
```
Comment (Erfan-Ahmadi) [inline examples_tests/05.NablaTutorialExample/main.cpp:243]: We don't need to call this twice @Przemog1, remove one.

**Pull request merged**